### DISCUSSION TITLE
Support for inputs on pcf8574

### DISCUSSION THREAD

**Discussion by turin400** [arendst/Tasmota#11176] [Ideas]
I have designed a pcb with 4 relay output, 4 input and pzem-16 485 support, I have compiled version 9.2 with the option for pcf8574 and I get everything to work except the inputs. I have configured them with the web interface but I don't have any message in the log when I change their status.
Could you help me.
Thanks.

**Comment by barbudor** [comment]
I think nothing has been implemented to manage PCF8574 inputs
I can't see anything in the code that deals with input
You should use a MCP230xx

**Reply by sfromis** [reply]
Or use ESP32, which has a lot more gpio pins....

**Comment by turin400** [comment]
I have read that the development of xdrv_28_pcf8574.ino by Stefan Bode is pending to add input as a button. I don't know if it's stopped.

**Reply by barbudor** [reply]
Not heard about it for months.

**Reply by sfromis** [reply]
Except for routine refactoring, no driver updates for the last year. By now, the word "pending" may overstate things a bit.....

**Reply by barbudor** [reply]
@turin400 Can you switch to MCP230xxx ?

Otherwise, what level of input support do you need ? Do you need full "Switch" equivalent with supports for Switchmodes etc ...
Or just a basic "sensor" type GPIO (SENSOR message and Rules trigger)

**Reply by sfromis** [reply]
If switching chip, better go to ESP32; easier to work with than a port extender.

**Reply by turin400** [reply]
@barbudor, I only need a basic "sensor" type GPIO (SENSOR message and Rules trigger). To change to MCP230xxx I would have to repeat the pcb and I have everything assembled.

**Reply by turin400** [reply]
![Captura](https://user-images.githubusercontent.com/42274403/109698897-3040eb00-7b90-11eb-8042-4eedf4aeb397.JPG)
I was very happy until I checked the pcf8574 problem with the inputs in tasmota.

**Reply by barbudor** [reply]
I will try to take some time this week-end for a simple input support that works with rules

**Reply by turin400** [reply]
Thanks for everything.

**Comment by barbudor** [comment]
@turin400
PR https://github.com/arendst/Tasmota/pull/11252 submitted
While it is under review, if you know how to pull a fork, you can try : https://github.com/barbudor/Tasmota/tree/pcf8574_inputs
Or I can build a specific firmware for you if you want

**Reply by barbudor** [reply]
Or pull Tasmota Development, replace `xdrv_28_pcf8574.ino` by attachement below
You will need some or all of the following in your `user_config_override.h`
```
      #define USE_PCF8574_SENSOR                   // enable PCF8574 inputs and outputs in SENSOR message
      #define USE_PCF8574_DISPLAYINPUT             // enable PCF8574 inputs in Web page
      #define USE_PCF8574_MQTTINPUT                // enable MQTT message on input change detection : stat/%topic%/PCF8574_INP = {"Time":"2021-03-07T16:19:23+01:00","PCF8574-1_INP":{"D1":1}}
      #define USE_PCF8574_EVENTINPUT               // enable Event generation on input change detection :  stat/%topic%/EVENT PCF8574-1_D1=1
```

[xdrv_28_pcf8574.ino.txt](https://github.com/arendst/Tasmota/files/6097585/xdrv_28_pcf8574.ino.txt) (rename ino.txt to .ino)

**Reply by barbudor** [reply]
Sensor message when `USE_PCF8574_SENSOR` is enabled:
```
18:13:26.462 MQT: tele/nodemcu/SENSOR = {"Time":"2021-03-07T18:13:26+01:00","PCF8574-1":{"D0":1,"D1":1,"D2":1,"D3":1,"D4":1,"D5":1,"D6":1,"D7":1}}
```

Web GUI when `USE_PCF8574_DISPLAYINPUT` is enabled (only inputs are listed, outputs are already shown as Power):
![image](https://user-images.githubusercontent.com/2996042/110248453-84254880-7f71-11eb-85be-615006680f4b.png)

Input change detection message when `USE_PCF8574_MQTTINPUT` is enabled:
```
18:16:11.475 MQT: stat/nodemcu/PCF8574_INP = {"Time":"2021-03-07T18:16:11+01:00","PCF8574-1_INP":{"D2":0}}
```
Mostly intended for processing by backend. For Rules see below:

Input change detection event when `USE_PCF8574_EVENTINPUT` is enabled: `EVENT PCF8574-1_D1=1`
Can be processed in a rule with (for any change):
```
ON event#PCF8574-1_D1 DO something %value% ENDON
```
or from a change from 0 to 1:
```
ON event#PCF8574-1_D1=1 DO something %value% ENDON
```

**Reply by turin400** [reply]
@barbudor 

Great job. I have tested it with my pcb and it works fine. Tb I have tested the rules and everything is fine.
If you need me to do any more tests, tell me.
Thanks again.

**Reply by barbudor** [reply]
Cool.
Can you post a note on the PR?
Note that may not be the final implementation depending on feedback from the team.

**Reply by tgreif** [reply]
hi i'm thomas , working on the same argumento , making a extension card 4 input + 4 output . 
the pcf8574 is working well as 8 output on my tasmota firmware . i've tryed to follow your indication , download the fork , override the driver file and make changes on my_user_config.h . also i've uncoment #define USE_PCF8574. 
its working ! 
perfect . 
can I switch directly the relè like a normal ESP9266 pin do .? 
can i use the output as shutter and input as sutter command? 
bye

.

**Reply by tgreif** [reply]
turin 800 please share your layout 
bye

**Comment by tgreif** [comment]
great job . 
i've followed your instruction and everything is working. 
it is possible to toggle output directly by input .

**Reply by barbudor** [reply]
You can use a rule based on event to toggle power
Beware there is no debouncing

**Reply by tgreif** [reply]
> 
> 
> You can use a rule based on event to toggle power
> Beware there is no debouncing

perhps its better to do by mqtt

**Comment by tgreif** [comment]
turin please share your layout 
thanks

**Reply by turin400** [reply]
it is a very basic design with kicad pcb. Hope this can help you.
[pasillo_10.zip](https://github.com/arendst/Tasmota/files/6108453/pasillo_10.zip)

**Comment by tgreif** [comment]
hi , 

#torino 400 grazie ! i will share my asap. 

#barboudor
i've tried to make rule . 
like this 
"event#PCF8574-1_D4=1 DO power4 1ENDON"
"event#PCF8574-1_D4=0 DO power4 0 ENDON"

if i trigger the input is sown the msg for input trigger and " "event"= "done". 
but it do not change relè state. 
i'm sure that i mistake but do not know where . is the first time on userevents. #
![grafik](https://user-images.githubusercontent.com/2463499/110474320-d1bfc380-80df-11eb-9da2-9c3b8e06c0a6.png)

**Reply by barbudor** [reply]
Can you post here the result of `Rule0`  (single word, no spaces) ?

**Reply by barbudor** [reply]
Working fine here:
```
16:59:03.202 MQT: stat/nodemcu/RULE = {"Rule1":"ON","Once":"OFF","StopOnError":"OFF","Length":44,"Free":467,"Rules":"on event#PCF8574-1_D0 do power1 toggle endon"}
16:59:06.691 MQT: stat/nodemcu/PCF8574_INP = {"Time":"2021-03-09T16:59:06+01:00","PCF8574-1_INP":{"D0":0}}
16:59:06.700 MQT: stat/nodemcu/EVENT = {"Event":"Done"}
16:59:06.743 RUL: EVENT#PCF8574-1_D0 performs "power1 toggle"
16:59:06.752 MQT: stat/nodemcu/POWER1 = {"POWER1":"ON"}
16:59:06.757 MQT: stat/nodemcu/POWER1 = ON (retained)
16:59:10.625 MQT: stat/nodemcu/PCF8574_INP = {"Time":"2021-03-09T16:59:10+01:00","PCF8574-1_INP":{"D0":1}}
16:59:10.635 MQT: stat/nodemcu/EVENT = {"Event":"Done"}
16:59:10.761 RUL: EVENT#PCF8574-1_D0 performs "power1 toggle"
16:59:10.768 MQT: stat/nodemcu/POWER1 = {"POWER1":"OFF"}
16:59:10.774 MQT: stat/nodemcu/POWER1 = OFF (retained)
```
You may have a rule syntax proble. Here is mine:
```
rule1 on event#PCF8574-1_D0 do power1 toggle endon
```
I'm using it as a button (any change => toggle) but you could use like a switch in follow-mode with
```
rule1 on event#PCF8574-1_D0 do power1 %value% endon
```

**Comment by tgreif** [comment]
#torino 400 

sotto il mio . non è ancora pronto perche mancano i ingressi 

![grafik](https://user-images.githubusercontent.com/2463499/110474546-18adb900-80e0-11eb-9711-5c0026492c83.png)

**Comment by tgreif** [comment]
hi thanks for your promt response , very kind.!

![grafik](https://user-images.githubusercontent.com/2463499/110481297-b3f65c80-80e7-11eb-9190-41bff4154f58.png)

**Reply by barbudor** [reply]
See my answer above
And please do not post screen shots but copy paste text with triple back ticks like:
![image](https://user-images.githubusercontent.com/2996042/110500592-a9dd5980-80f9-11eb-8feb-0acdb998a127.png)

**Comment by tgreif** [comment]
Hi , 
it ' worrking thanks for help . like you told only momentary . 

bye

**Comment by barbudor** [comment]
TO ALL : The PR is still in review by Theo.
Depending on Theo's decision the final may be different.
Thank you for testing and sharing your result but do not consider it as definitive although I do not expect it to be rejected, there may be some changes.

**Reply by turin400** [reply]
Until now there was no support for input. It's a good start @barbudor.

**Comment by barbudor** [comment]
Final PR : https://github.com/arendst/Tasmota/pull/11301

We have removed the EVENT.
Rules can process Input change using USE_PCF8574_MQTTINPUT and
`on PCF8574-1_INP#D0 do ...` or `on PCF8574-1_INP#D0==1 do ...`

**Comment by Jason2866** [comment]
PR is merged. Thx @barbudor

**Comment by jacekmarek** [comment]


**Reply by sfromis** [reply]
From the perspective of Tasmota, PCF8574 pins are not buttons, merely inputs which can change state. Since your device is ESP32-based, with Berry you have a good programming environment for coding additional handling, like detecting multiple actions within a short time.

A "real" button is what's configured on gpio 0.
